flang/lib/Lower/ReductionRuntime.cpp

```diff
@@ -229,10 +344,12 @@ void Fortran::lower::genMaxlocDim(Fortran::lower::FirOpBuilder &builder,
 }
 
 /// Generate call to Maxval intrinsic runtime routine. This is the version
-/// that does not take a dim argument.
+/// that does not take a dim argument. Note: resultBox is not currently used
+/// but is needed so we can share a template with SUM and PRODUCT intrinsics.
 mlir::Value Fortran::lower::genMaxval(Fortran::lower::FirOpBuilder &builder,
                                       mlir::Location loc, mlir::Value arrayBox,
-                                      mlir::Value maskBox) {
+                                      mlir::Value maskBox,
+                                      mlir::Value resultBox) {
```

Comment:
In my opinion, it adds more complexity to try sharing `genExtremumValProdSum` by having this unused `resultBox` argument here rather than having too functions `genExtremumVal` and `genProdOrSum`.

Reply:
Agreed. It's increasing the cyclomatic complexity (extra `if` statements being salted in here, there, etc.). It makes it harder to understand which conditions are valid for which intrinsic(s) and invalid in the rest. Or possibly if some error-by-usage cases are simply unchecked.